### ISSUE TITLE
How can we stop training if there is not eval. set improvement?

### ISSUE THREAD

**Issue by erogol** [dmlc/xgboost#243]
As the training goes, I see the eval results but I cannot see any scheduling to terminate learning when eval values stop to improve. Is there any such mechanism. If there is not, I think it is a good addition

**Comment by tqchen** [comment]
https://github.com/dmlc/xgboost/wiki/Using-XGBoost-in-Python#early-stopping

**Comment by achm6174** [comment]
Hello,
Are there any equivalent setting in R for the cross validation function? 
Thanks.

**Comment by hetong007** [comment]
@achm6174 This feature is added. The demo is available here:[https://github.com/dmlc/xgboost/blob/master/R-package/demo/early_Stopping.R#L38](https://github.com/dmlc/xgboost/blob/master/R-package/demo/early_Stopping.R#L38)

**Comment by achm6174** [comment]
@hetong007 Thanks this feature definitively save me lots of time in training, I'll try later.
i will keep seeing if there's anything i can contribute/ fix or bug report about the R library.

**Comment by hetong007** [comment]
@achm6174 How's going with this feature? Does it work?

**Comment by achm6174** [comment]
@hetong007 
It works but I have a question when I am trying to implement this on a real dataset, which is a regression problem.
I ran CV but apparently the early stopping was based on the train-rmse instead of the test-rmse. Here's the output:

    load("benchmark.RData")

    param
    #$objective
    #[1] "reg:linear"
    #
    #$max_depth
    #[1] 7
    # 
    # $eta
    # [1] 0.1
    # 
    # $min_child_weight
    # [1] 2
    # 
    # $max_delta_step
    # [1] 0
    # 
    # $subsample
    # [1] 0.4
    # 
    # $colsample_bytree
    # [1] 0.4

    set.seed(6174)
    cv.nround = 2000
    bst.cv = xgb.cv(param=param, data = x, label = y,
    nfold = 10, nrounds=cv.nround, maximize = FALSE, early.stop.round = 2)

    # [349] train-rmse:0.062263+0.016703 test-rmse:0.422103+0.086629
    # [350] train-rmse:0.062152+0.016708 test-rmse:0.421890+0.086773
    # [351] train-rmse:0.061987+0.016864 test-rmse:0.421531+0.086790
    # [352] train-rmse:0.061984+0.016903 test-rmse:0.421153+0.086654
    # [353] train-rmse:0.061945+0.016971 test-rmse:0.421265+0.086831
    # [354] train-rmse:0.061871+0.017078 test-rmse:0.421256+0.086740
    # [355] train-rmse:0.061734+0.017058 test-rmse:0.421029+0.086624
    # [356] train-rmse:0.061776+0.017087 test-rmse:0.420698+0.086241
    # [357] train-rmse:0.061661+0.017122 test-rmse:0.420673+0.086210
    # [358] train-rmse:0.061511+0.017174 test-rmse:0.420213+0.086219
    # [359] train-rmse:0.061416+0.017204 test-rmse:0.420147+0.085853
    # [360] train-rmse:0.061413+0.017236 test-rmse:0.420161+0.085790
    # [361] train-rmse:0.061389+0.017347 test-rmse:0.420042+0.085868
    # [362] train-rmse:0.061253+0.017340 test-rmse:0.419980+0.085610
    # [363] train-rmse:0.061050+0.017353 test-rmse:0.419947+0.085641
    # [364] train-rmse:0.060960+0.017489 test-rmse:0.420040+0.085710
    # [365] train-rmse:0.060878+0.017536 test-rmse:0.420042+0.085498
    # [366] train-rmse:0.060833+0.017605 test-rmse:0.419703+0.085272
    # [367] train-rmse:0.060777+0.017662 test-rmse:0.419524+0.085277
    # [368] train-rmse:0.060617+0.017722 test-rmse:0.420279+0.084461
    # [369] train-rmse:0.060514+0.017688 test-rmse:0.420183+0.084550
    # [370] train-rmse:0.060356+0.017760 test-rmse:0.420270+0.084494
    # [371] train-rmse:0.060298+0.017861 test-rmse:0.420410+0.084637
    # [372] train-rmse:0.060245+0.017885 test-rmse:0.420491+0.084650
    # [373] train-rmse:0.060167+0.017917 test-rmse:0.420381+0.085066
    # [374] train-rmse:0.060097+0.017930 test-rmse:0.420455+0.084983
    # [375] train-rmse:0.059958+0.017890 test-rmse:0.420430+0.085015
    # [376] train-rmse:0.059827+0.017902 test-rmse:0.420441+0.085303
    # [377] train-rmse:0.059823+0.017921 test-rmse:0.420338+0.085265
    # [378] train-rmse:0.059748+0.017921 test-rmse:0.420380+0.084981
    # [379] train-rmse:0.059644+0.017893 test-rmse:0.420335+0.084848
    # [380] train-rmse:0.059490+0.018016 test-rmse:0.420347+0.084421
    # [381] train-rmse:0.059457+0.018051 test-rmse:0.420493+0.084454
    # [382] train-rmse:0.059372+0.018087 test-rmse:0.421065+0.085066
    # [383] train-rmse:0.059264+0.018119 test-rmse:0.421052+0.084796
    # [384] train-rmse:0.059193+0.018161 test-rmse:0.420957+0.084687
    # [385] train-rmse:0.059097+0.018206 test-rmse:0.421025+0.084571
    # [386] train-rmse:0.059068+0.018328 test-rmse:0.420825+0.084572
    # [387] train-rmse:0.059002+0.018360 test-rmse:0.420744+0.085051
    # [388] train-rmse:0.058910+0.018377 test-rmse:0.420916+0.085109
    # [389] train-rmse:0.058888+0.018421 test-rmse:0.421006+0.085146
    # [390] train-rmse:0.058850+0.018462 test-rmse:0.421026+0.085229
    # [391] train-rmse:0.058772+0.018477 test-rmse:0.421083+0.085118
    # [392] train-rmse:0.058700+0.018480 test-rmse:0.421037+0.085502
    # [393] train-rmse:0.058648+0.018502 test-rmse:0.421081+0.085504
    # [394] train-rmse:0.058594+0.018513 test-rmse:0.421262+0.085522
    # [395] train-rmse:0.058597+0.018576 test-rmse:0.421735+0.086049
    # [396] train-rmse:0.058606+0.018649 test-rmse:0.421770+0.086094
    # Stopping. Best iteration: 395

**Comment by hetong007** [comment]
@achm6174 Thanks for catching that! I just pushed a simple fix.

**Comment by LancelotHolmes** [comment]
I'm confused about when to use the early_stopping, say if my pipeline is like:
* k-fold cv to tune the model params 
* use all training data to train the model 
* finally predict on the test set

 my question is when should we use early_stopping, cv stage or training stage?

if I use the early_stopping in cv stage, then how can I choose the best_ntree_limit or best iteration on my training stage, or should I separate my training set to use like 20%  training data to eval and get the best_ntree_limit which can be used on the prediction? 
for example:
* first  use cv to select model or tune model params
* then split the train set like(80%-20%) and use early-stopping to train the model and get the best_ntree_limit 
* finally use the model return from train stage to predict on the test set with the best_ntree_limit 
Is there repetitive work? or what's the correct workflow?
